### PR TITLE
HQLPARSER-65 Add support for translating GROUP BY and HAVING clauses

### PR DESCRIPTION
**Pull request by anistor**
jira: https://hibernate.atlassian.net/browse/HQLPARSER-65

### CODE REVIEW

File: parser/src/main/java/org/hibernate/hql/ast/origin/hql/resolve/path/PropertyPath.java [hunk 1]
```diff
@@ -27,14 +27,31 @@
...
- * e.g. {@code foo.bar.baz}.
+ * Represents a path of properties (e.g. {@code foo.bar.baz}) represented by {@link PathedPropertyReferenceSource}s used
+ * in a SELECT, GROUP BY, WHERE or HAVING clause.
  *
  * @author Gunnar Morling
  */
 public class PropertyPath {
 
-	private final LinkedList<PathedPropertyReferenceSource> path = new LinkedList<PathedPropertyReferenceSource>();
+	protected final LinkedList<PathedPropertyReferenceSource> path;
```
Comment (gunnarmorling) [inline parser/src/main/java/org/hibernate/hql/ast/origin/hql/resolve/path/PropertyPath.java]: Why does it have to be `protected`?

File: parser/src/main/java/org/hibernate/hql/ast/spi/SingleEntityQueryRendererDelegate.java [hunk 2]
```diff
@@ -288,7 +430,19 @@ public void sortSpecification(String collateName, String orderSpec) {
     * @param collateName optional collation name
     * @param isAscending sort direction
     */
-	protected abstract void addSortField(PropertyPath propertyPath, String collateName, boolean isAscending);
```
Comment (gunnarmorling) [inline parser/src/main/java/org/hibernate/hql/ast/spi/SingleEntityQueryRendererDelegate.java]: What's the rationale behind this change from `abstract`?

**Comment by gunnarmorling** [comment]
@anistor Looks alright to me apart from the two questions above. Feel free to proceed.

File: parser/src/main/java/org/hibernate/hql/ast/origin/hql/resolve/path/PropertyPath.java (hunk 1 above)
Comment (anistor) [inline parser/src/main/java/org/hibernate/hql/ast/origin/hql/resolve/path/PropertyPath.java]: It is used by subclass (AggregationPropertyPath.toString())

File: parser/src/main/java/org/hibernate/hql/ast/spi/SingleEntityQueryRendererDelegate.java (hunk 2 above)
Comment (anistor) [inline parser/src/main/java/org/hibernate/hql/ast/spi/SingleEntityQueryRendererDelegate.java]: Implementing sorting and grouping should be optional but should not require users to provide any implementation not even a blank method if they do not wish to support it. This change is backward compatible.

**Comment by anistor** [comment]
Thanks a lot @gunnarmorling !  I've just made a minor update to some javadoc. 
Are you ok to merge this ? Infinispan also needs a HQL release (I can handle that myself).

**Comment by anistor** [comment]
Integrated in master.
